### PR TITLE
updates the local API service to include CORS headers

### PR DESCRIPTION
**Pull request by jlapier**
Previous to this commit, `sam local api` would only return CORS headers when an OPTIONS request is made. This causes issues for browsers, which expect at the very least "Access-Control-Allow-Origin" to be present on other HTTP methods. This commit adds the "Access-Control-Allow-Origin" header to a request if it is made with one of the Allowed-Methods.

References:
 * Fetch spec on CORS: https://fetch.spec.whatwg.org/#cors-request
 * Examples from Mozilla: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
 * API Gateway doc on when to supply CORS headers: https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html

#### Which issue(s) does this change fix?

#1460 

#### Why is this change necessary?

In order to test a local UI against a local API via sam, proper CORS headers are required or the browser will reject the requests.

#### How does it address the issue?

Adds the Access-Control-Allow-Origin header if a CORS configuration is present for the API.

#### What side effects does this change have?

The CORS header behavior is not entirely predictable for REST APIs, according to the [AWS API Gateway docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html). This change may mask a need to include the headers at the integration point. However, for [HTTP APIs](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-cors.html), the behavior is much more predicable and this change is consistent with that behavior.

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by vmarinescu — APPROVED** [review]

File: samcli/local/apigw/local_apigw_service.py [hunk 1]
```diff
@@ -375,8 +375,20 @@ def _request_handler(self, **kwargs):
...
         return self.service_response(body, headers, status_code)
 
+    @staticmethod
+    def _add_cors_headers_to_response(method, response_headers, cors_headers):
+        # If the CORS headers are present and one of the allowed methods is being used,
+        # add the allow-origin header to the response.
+        # On Origin other than the wildcard (*), include Vary: Origin
+        allow_origin = cors_headers.get("Access-Control-Allow-Origin")
+        if method in cors_headers.get("Access-Control-Allow-Methods", []) and allow_origin:
+            response_headers.add_header("Access-Control-Allow-Origin", allow_origin)
```
Comment (jfuss) [inline samcli/local/apigw/local_apigw_service.py:388]: According to the API Gateway docs, the Lambda function is responsible for adding these integrations: https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html (in the "Enabling CORS support for Lambda or HTTP proxy integrations" section). So SAM CLI should not be doing this.

File: samcli/local/apigw/local_apigw_service.py [hunk 2]
```diff
@@ -375,8 +375,20 @@ def _request_handler(self, **kwargs):
...
+            if allow_origin != "*":
+                response_headers.add_header("Vary", "Origin")
```
Comment (jfuss) [inline samcli/local/apigw/local_apigw_service.py:390]: Does API Gateway add this?

File: samcli/local/apigw/local_apigw_service.py [hunk 3]
```diff
@@ -768,7 +780,7 @@ def _construct_v_2_0_event_http(
             # Flask does not parse/decode the request data. We should do it ourselves
             request_data = request_data.decode("utf-8")
 
-        query_string_dict, _ = LocalApigwService._query_string_params(flask_request)
+        query_string_dict = LocalApigwService._query_string_params_v_2_0(flask_request)
```
Comment (jfuss) [inline samcli/local/apigw/local_apigw_service.py:783]: I don't understand how this relates to the CORS headers.

**Review by jfuss — CHANGES_REQUESTED** [review]

**Comment by jfuss** [comment]
Closing given no response. Feel free to address comments and re-open and we will re-review.

**Pull request closed without merging**